File: python/leetcode/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        self.paths = []
        self.findSum(root, 0, sum, [])
        return self.paths
        
    def findSum(self, root, this_sum, sum, this_path):
        if root is None:
            return
        new_sum = this_sum + root.val
        this_path.append(root.val)
        if root.left is None and root.right is None:
            if sum == new_sum:
                self.paths.append(this_path[:])
        else:
            if root.left is not None:
                self.findSum(root.left, new_sum, sum, this_path)
            if root.right is not None:
                self.findSum(root.right, new_sum, sum, this_path)
        this_path.pop()

File: java/leetcode/path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<List<Integer>> paths;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        paths = new ArrayList();
        List<Integer> thisPath = new ArrayList();
        findSum(root, 0, sum, thisPath);
        return paths;
    }
    
    void findSum(TreeNode root, int thisSum, int sum, List<Integer> thisPath) {
            if (root == null) { return; }
            thisSum += root.val;
            thisPath.add(root.val);
            if (root.left == null && root.right == null) {
                if (sum == thisSum) {
                    paths.add(new ArrayList(thisPath));
                }
            } else {
                if (root.left != null) {
                    findSum(root.left, thisSum, sum, thisPath);
                }
                if (root.right != null) {
                    findSum(root.right, thisSum, sum, thisPath);
                }
            }
            thisSum -= root.val;
            thisPath.remove(thisPath.size() - 1);
    }
}
